File: solutions/python3/0105.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        def helper(pLeft: int, pRight: int, iLeft: int, iRight: int) -> TreeNode:
            if pLeft > pRight or iLeft > iRight:
                return None

            i = dict[preorder[pLeft]]
            curr = TreeNode(preorder[pLeft])
            curr.left = helper(pLeft + 1, pLeft + i - iLeft, iLeft, i - 1)
            curr.right = helper(pLeft + i - iLeft + 1, pRight, i + 1, iRight)

            return curr

        dict = {val: i for i, val in enumerate(inorder)}

        return helper(0, len(preorder) - 1, 0, len(inorder) - 1)

File: solutions/cpp/0105.cpp
class Solution {
 public:
  TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
    unordered_map<int, int> map;
    for (int i = 0; i < inorder.size(); ++i) map[inorder[i]] = i;

    return helper(preorder, 0, preorder.size() - 1, inorder, 0,
                  inorder.size() - 1, map);
  }

  TreeNode* helper(vector<int>& preorder, int pLeft, int pRight,
                   vector<int>& inorder, int iLeft, int iRight,
                   unordered_map<int, int>& map) {
    if (pLeft > pRight || iLeft > iRight) return nullptr;

    int i = map[preorder[pLeft]];
    TreeNode* curr = new TreeNode(preorder[pLeft]);
    curr->left = helper(preorder, pLeft + 1, pLeft + i - iLeft, inorder, iLeft,
                        i - 1, map);
    curr->right = helper(preorder, pLeft + i - iLeft + 1, pRight, inorder,
                         i + 1, iRight, map);

    return curr;
  }
};